Mage.Sets/src/mage/cards/r/RainOfRiches.java

```diff
@@ -127,19 +130,30 @@ class RainOfRichesWatcher extends Watcher {
 
     @Override
     public void watch(GameEvent event, Game game) {
-        if (event.getType() != GameEvent.EventType.CAST_SPELL) {
+        if (event.getType() != GameEvent.EventType.MANA_PAID) {
             return;
         }
-        Spell spell = game.getSpell(event.getSourceId());
-        if (spell == null) {
+        ManaPaidEvent manaEvent = (ManaPaidEvent) event;
+
+        if (!manaEvent.getSourceObject().hasSubtype(SubType.TREASURE, game)) {
             return;
         }
-        int manaPaid = ManaPaidSourceWatcher.getTreasurePaid(spell.getId(), game);
-        if (manaPaid < 1) {
+
+        Optional<Ability> opt = game.getAbility(manaEvent.getTargetId(), manaEvent.getSourcePaidId());
```

Comment:
this seems unnecessarily verbose. Already checking `instanceof` later on which takes care of the null check so I don't see the point in the optional.

Reply:
Because `.get()` throws an exception if there isn't an ability - it doesn't return null.